### ISSUE TITLE
How to perform inference for YOLONAS with quantization exported using NMS for TensorRT?

### ISSUE THREAD

**Issue by NurAlif** [Deci-AI/super-gradients#1451]
### 💡 Your Question

I successfuly train and perform quantization using `Trainer.quantize_from_config(cfg)`.

Then I tried export it for both `ONNXRUNTIME` and `TENSORRT`:
```
model_qat.export(
    output= "experiment1/yolo_nas_s_nms_onnx.onnx", 
    input_image_shape=(320,480),
    quantization_mode=ExportQuantizationMode.INT8,
    calibration_loader=train_loader,
    engine=ExportTargetBackend.ONNXRUNTIME
    )
```  
```
model_qat.export(
    output= "experiment1/yolo_nas_s_nms_trt.onnx", 
    input_image_shape=(320,480),
    quantization_mode=ExportQuantizationMode.INT8,
    calibration_loader=train_loader,
    engine=ExportTargetBackend.TENSORRT
    )
```  

The first model with `onnxruntime` target was working correctly. I confirm it by drawing bounding boxes from the model output on top of the image, and its correct. However, the second model with `tensorrt` target doesn't seem to work properly. The second model converted to `.trt` using this command: `trtexec --onnx=experiment1/yolo_nas_s_nms_trt.onnx --saveEngine=yolonas_qtz_nms.trt --int8`. Then I test it using the following code:
```
w = 'yolonas_qtz_nms.trt'
device = torch.device('cuda:0')

bindings = None
binding_addrs = None
context = None

color_ball = (0, 0, 255)
color_goal = (255, 0, 0)


# Infer TensorRT Engine
def init_tensorrt():
    global bindings
    global binding_addrs
    global context

    Binding = namedtuple('Binding', ('name', 'dtype', 'shape', 'data', 'ptr'))
    logger = trt.Logger(trt.Logger.INFO)
    trt.init_libnvinfer_plugins(logger, namespace="")
    with open(w, 'rb') as f, trt.Runtime(logger) as runtime:
        model = runtime.deserialize_cuda_engine(f.read())
    bindings = OrderedDict()
    for index in range(model.num_bindings):
        name = model.get_tensor_name(index)
        dtype = trt.nptype(model.get_tensor_dtype(name))
        shape = tuple(model.get_tensor_shape(name))
        data = torch.from_numpy(np.empty(shape, dtype=np.dtype(dtype))).to(device)
        bindings[name] = Binding(name, dtype, shape, data, int(data.data_ptr()))
    binding_addrs = OrderedDict((n, d.ptr) for n, d in bindings.items())
    context = model.create_execution_context()

        # warmup for 10 times
    for _ in range(10):
        tmp = torch.randn(1,3,320,480).to(device)
        binding_addrs['onnx::Cast_0'] = int(tmp.data_ptr())
        context.execute_v2(list(binding_addrs.values()))

def detect():

    global bindings
    global binding_addrs
    global context
    global videocap

    img = videocap.read_in()

    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    image = img.copy()
    image = image.transpose((2, 0, 1))
    image = np.expand_dims(image, 0)

    im = image.astype(np.uint8)
    im = torch.from_numpy(im).to(device)

    start = time.perf_counter()
    binding_addrs['onnx::Cast_0'] = int(im.data_ptr())
    context.execute_v2(list(binding_addrs.values()))
    exec_cost = time.perf_counter()-start

    nums = bindings['num_dets'].data
    boxes = bindings['det_boxes'].data
    classes = bindings['det_classes'].data

    boxes = boxes[0,:nums[0][0]]
    classes = classes[0,:nums[0][0]]

    for box,cl in zip(boxes,classes):
        box = box.round().int().tolist()
        print(box)
        cv2.rectangle(img,(box[0],box[1]),(box[2],box[3]),color_goal,2)
```

The result of the bounding boxes positions are random, but the size of the bounding boxes is kind of working but still not correct:
![image](https://github.com/Deci-AI/super-gradients/assets/21985252/44202c98-df90-4a78-a264-41d82f384475)
The blue ball and white pole is the object detection target. The whole predicted bounding boxes on every frame of the video is not correct.

I also checked my `yolo_nas_s_nms_trt.onnx` with netron. Indeed it has `EfficientNMS_TRT` node:
![image](https://github.com/Deci-AI/super-gradients/assets/21985252/86471e51-5559-498c-8383-85e9328c446b)

I have not found any documentation on how to specifically export quantized YOLONAS with NMS for TensorRT and perform inference on the model. I still have no idea is that my inference code that is wrong or it is has something to do with the `model.export()`.

### Question:
1. Do I need to add `quantization_mode=ExportQuantizationMode.INT8` on `model.export()` after performing quantization on the model?
2. Can anyone who has tried exporting YOLONAS with quantization and NMS using `model.export()` confirm that it can correctly produce working model for tensorrt engine?

### Versions

**JETSON ORIN NANO JP5.1.1**

PyTorch version: 2.0.0+nv23.05
Is debug build: False
CUDA used to build PyTorch: 11.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (aarch64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 8.0.1-9 (tags/RELEASE_801/final)
CMake version: version 3.27.2
Libc version: glibc-2.31

Python version: 3.8.10 (default, May 26 2023, 14:05:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.10.104-tegra-aarch64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.4.315
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/aarch64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/aarch64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/aarch64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/aarch64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/aarch64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/aarch64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/aarch64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                    aarch64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
CPU(s):                          6
On-line CPU(s) list:             0-5
Thread(s) per core:              1
Core(s) per socket:              6
Socket(s):                       1
Vendor ID:                       ARM
Model:                           1
Model name:                      ARMv8 Processor rev 1 (v8l)
Stepping:                        r0p1
CPU max MHz:                     1510,4000
CPU min MHz:                     115,2000
BogoMIPS:                        62.50
L1d cache:                       384 KiB
L1i cache:                       384 KiB
L2 cache:                        1,5 MiB
L3 cache:                        2 MiB
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; __user pointer sanitization
Vulnerability Spectre v2:        Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp uscat ilrcpc flagm

Versions of relevant libraries:
[pip3] numpy==1.23.0
[pip3] pytorch-quantization==2.1.2
[pip3] torch==2.0.0+nv23.5
[pip3] torchmetrics==0.8.0
[pip3] torchvision==0.15.2

**Comment by BloodAxe** [comment]
Thanks for the bugreport.

Have you tried exporting a regular model (FP32, ONNX) and running it through onnxruntime as a sanity check?
I'm trying to bisect where the problem is originating. Is it the model itself, quantization or TRT.

We tested whether exported models (FP32,FP16 & INT8) works under onnxruntime. On TRT visual checks was not performed, only benchmarking was done there. 

Here you can see an example what was tested:
* https://github.com/Deci-AI/super-gradients/blob/master/src/super_gradients/examples/model_export/models_export.ipynb
* https://docs.deci.ai/super-gradients/latest/documentation/source/models_export.html#output-format-for-detection-models

So here is what I suggest you to do:
0) Try model.predict on the example image to see whether regular model gives you expected output
1) Export model with FP32 and see whether in produce expected output using onnxruntime
2) Export model with INT8 and see whether in produce expected output using onnxruntime

Pls keep us posted of your findings :)

**Comment by NurAlif** [comment]
0. I did `model.predict` on the unquantized model, and it is working as expected. Bounding boxes were drawn correctly.
1. Export `model_qat.export()` as FP32 model also working as expected. Bounding boxes from `onnxruntime` inference were drawn correctly.
2. As I did above, the quantized INT8 model with NMS using `onnxruntime` also working as expected. Bounding boxes from `onnxruntime` inference were drawn correctly.

I think the problem is related to the TRT NMS exporting, considering if the inference code above is correct.

I also want to mention this is how I preprocess the image before feeding it into the TRT engine:
````
    img = videocap.read_in()

    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    image = img.copy()
    image = image.transpose((2, 0, 1))
    image = np.expand_dims(image, 0)

    im = image.astype(np.uint8)
````
I did the same preprocessing as I tested INT8 model with NMS using `onnxruntime`. It is working correctly, so I consider the preprocessing to be correct.
